### DISCUSSION TITLE
TWIG Invoice: Iterate over Projects: Example please

### DISCUSSION THREAD

**Discussion by labor4** [kimai/kimai#3875] [Questions & Answers]
Hi all

I've gone through the twig examples at [DOCS PAGE](https://www.kimai.org/documentation/invoices.html), but I'm struggling to find a working solution.

What is the suggested way to **iterate over multiple project entries** in a twig template?
Would somebody have a simple example of the ``for`` structure, please?

The aim is to play out a PDF using the twig renderer, that can handle multiple project-selections at once.

Thank you very much

**Comment by kevinpapst** [comment]
> but I'm struggling to find a working solution.

share what you did

the trick is to merge the array in twig, something like that (untested)

{% set sheets = {} %}
{% for record in entries %}
{% set sheets = sheets|merge({(record.project.id): record}) %} ...

**Reply by labor4** [reply]
from my testbench:
input is 2 unique projects, from 3 entries total.

````
        {% set collectp = {} %}
        {% for thisentry in entries %}
            {% set collectp = collectp|merge({(thisentry['entry.project_id']): thisentry }) %} 
        {% endfor %}
        
        {% for thisid, check in collectp %}
        <tr class="myhead">
            <td class="">---------- here -----------</td>
            <td class="">{{ thisid }}</td>
        </tr>
        {% endfor %}

````

ouput is: 3 entries, no sub-array created.
the depth is the same as the original.

**Reply by labor4** [reply]
it should output 2x entries with ```thisid``` being the ```project_id```

**Reply by labor4** [reply]
is it correct there is nothing like:
````
$collect[$thisentry['entry.project_id']][$thisentry['entry.begin_timestamp']] = $thisentry;
````
in twig?
I think I still haven't got the structure right.
is there something like dump() to get an overview?

**Reply by labor4** [reply]
btw: https://www.branchcms.com/learn/docs/developer/twig/filters/item/merge-deep

**Reply by kevinpapst** [reply]
Last one does not exist in Kimai. 
Check the debug template at https://www.kimai.org/documentation/invoices.html#twig-templates

And try
             {% set collectp = collectp|merge({('project_' ~ thisentry['entry.project_id']): thisentry }) %}

**Reply by labor4** [reply]
your added string did magic.
however: entries get overwritten I believe.

therefore... I tried to work around that.
expecting i'll have to group them into another subgroup.
firstly:
```
        {% set collectp = {} %}
        {% set tempcoll = {} %}
        {% set thistime = 0 %}
        {% set count = 0 %}
        {% for thisentry in entries %}
            
            {% set count = count + 1 %}
            {% set collectp = collectp|merge({('project_' ~ thisentry['entry.project_id']): {('count_' ~ count) : thisentry } }) %}
            
        {% endfor %}
```
however: no, I still get 2 projects in total, holding 2 entries in total (should be 3).

then,
grab a chair on this one:
```
        {% set collectp = {} %}
        {% set tempcoll = {} %}
        {% set thistime = 0 %}
        {% for thisentry in entries %}
            
            
            {% for pairA, pairB in thisentry %}
            
                {% if pairA == "entry.begin_timestamp" %}
                    {% set thistime = pairB %}
                {% endif %}
                
                {% set tempcoll = tempcoll|merge({(pairA) : pairB }) %}
                
                {% if pairA == "entry.row" %}
                    
                    {% if thistime != 0 %}
                        {% set collectp = collectp|merge({('project_' ~ thisentry['entry.project_id']): {('unix_' ~ thistime) : tempcoll } }) %}
                    {% endif %}
                    
                    {% set tempcoll = {} %}
                    
                {% endif %}
                
            {% endfor %}
            
        {% endfor %}

```
what I did here is expecting every entry to have a unix timestamp (so three in total).
going into the for loop and making notes, adding to master merge, upon row "entry.row", with an exclusion.

the depth is correct. I get
```
projectid =>
    unix_TIMESTAMP =>
        ARRAYOFVALUES
```
however: no, I still get 2 projects in total, holding 2 entries in total.
so still one gets overwritten.
Couldn't PHP prepare these things much better?

**Reply by kevinpapst** [reply]
I think the issue is that Twig does not merge recursively.
You need to do two merges, first the array of timesheets and then the global array, something like:
```
            {% set id = 'project_' ~ thisentry['entry.project_id'] %}
            {% set entries = collectp[id]|merge({(thisentry['entry.id']): thisentry})
            {% set collectp = collectp|merge({(id): entries}) %}
```

**Comment by labor4** [comment]
```
        {% set collectp = {} %}
        {% set thistime = 0 %}
        {% set count = 0 %}
        {% for thisentry in entries %}
        
            {% set collentries = {} %}
            {% set count = count + 1 %}
            
            {% set id = 'project_' ~ thisentry['entry.project_id'] %}
            {% set entries = collectp[id]|merge({(thisentry['entry.id']): thisentry})
            {% set collectp = collectp|merge({(id): entries}) %}
            
        {% endfor %}
```
makes:
```Änderungen konnten nicht gespeichert werden: Unexpected character "!".```

and:
```        {% set collectp = {} %}
        {% set thistime = 0 %}
        {% set count = 0 %}
        {% for thisentry in entries %}
        
            {% set collentries = {} %}
            {% set count = count + 1 %}
            
            {% set id = 'project_' ~ thisentry['entry.project_id'] %}
            {% set collentries = collentries[id]|merge({(thisentry['entry.id']): thisentry})
            {% set collectp = collectp|merge({(id): collentries}) %}
            
        {% endfor %}
```
makes aswell
```Änderungen konnten nicht gespeichert werden: Unexpected character "!".```

is ```collentries[id] ``` good?

**Reply by kevinpapst** [reply]
I haven't tested the code and it is friday evening. I am sharing ideas so you can solve your issue. 
You can ask any development community about deep merging arrays in twig ... 
I know it works and I know code like that is in Kimai, just can't remember where.

Try `collentries[(id)]` - twig is really weird sometimes.

**Reply by labor4** [reply]
alright, no stress. I'm just very new to twig.
Thank you
Gutes Wochenende!

**Comment by labor4** [comment]
according to [this saviour](https://stackoverflow.com/a/72405741/3802870),
reordering of projects into a sub-array goes along these lines:
```
        {% set output = [] %}
        {% for item in entries %}
            {% set projstring = 'project_' ~ item['entry.project_id'] %}

            {% if not attribute(output, projstring) is defined %}
                {% set output = output|merge({ (projstring) : {}, }) %}
            {% endif %}

            {% set output = output|merge({(projstring) : output[projstring] | merge([ item, ]) }) %}
        {% endfor %}


```
